(SERVER-1103) Fix borrow-after-borrow lock tests

camlow325 opened the pull request:
This commit includes fixes for timing-related issues in the
`pool-can-borrow-after-borrow-interrupted-during-lock` and
`pool-can-borrow-after-borrow-timed-out-during-lock` tests.  In both
cases, there previously was a race condition where a borrow on a
separate thread could succeed before a lock() call was made, causing the
lock() call to block indefinitely because the pool was indefinitely not
full.  The production code was behaving properly but the tests were
designed with the expectation that the lock would occur before the
any borrows were attempted.  Both tests were changed to wait to perform
the borrow until the lock is taken.

camlow325 commented on test/unit/puppetlabs/puppetserver/lockable_pool_test.clj:
```diff
@@ -508,17 +508,19 @@
-          lock-thread-locked? (promise)
           unlock? (promise)
           lock-thread (future
                        (deliver lock-thread-started? true)
```
For both of these tests, I was able to reproduce the failures consistently on my local box by inserting a long sleep between the `(deliver lock-thread-started? true)` call and the `(.lock pool)` call.  I used a 9 second sleep - long enough to reliably cause the other `borrow-thread-*` futures to be able to make a `.borrowItem` calls - and for one to succeed after the `.releaseItem` call for the initially borrowed instance was made.  Less than 10 seconds was important to use in order to avoid hitting the "global" `timed-deref` timeout of 10 seconds that these tests arbitrarily use.

cprice404 replied:
So, just to clarify, for the semantics of this test, it is important that for threads 2 and 3, they do not attempt to borrow until after the lock is held?

camlow325 replied:
Yep, that's right.  They specifically are intending to validate what happens around borrow attempts that are made after the lock is already held.

camlow325 commented:
Travis failed on a second run with the following error:

~~~~
lein test :only puppetlabs.services.jruby.jruby-puppet-service-test/test-error-during-init

ERROR in (test-error-during-init) (jruby_puppet_agents.clj:60)
Uncaught exception, not in assertion.
expected: nil

  actual: java.lang.IllegalStateException: There was a problem adding a JRubyPuppet instance to the pool.
 at puppetlabs.services.jruby.jruby_puppet_agents$eval7581$prime_pool_BANG___7586$fn__7590.invoke (jruby_puppet_agents.clj:60)
    puppetlabs.services.jruby.jruby_puppet_agents$eval7581$prime_pool_BANG___7586.invoke (jruby_puppet_agents.clj:38)
puppetlabs.services.jruby.jruby_puppet_agents$eval7779$send_prime_pool_BANG___7784$fn__7785$fn__7787.invoke (jruby_puppet_agents.clj:181)
    puppetlabs.trapperkeeper.internal$shutdown_on_error_STAR_.invoke (internal.clj:254)
    puppetlabs.trapperkeeper.internal$shutdown_on_error_STAR_.invoke (internal.clj:238)
puppetlabs.trapperkeeper.internal$shutdown_service$reify__6433$service_fnk__5270__auto___positional$reify__6438.shutdown_on_error (internal.clj:293)
    puppetlabs.trapperkeeper.internal$eval6369$fn__6381$G__6361__6389.invoke (internal.clj:262)
    puppetlabs.trapperkeeper.internal$eval6369$fn__6381$G__6360__6398.invoke (internal.clj:262)
    clojure.core$partial$fn__4527.invoke (core.clj:2494)
    clojure.core$partial$fn__4527.invoke (core.clj:2493)
    puppetlabs.services.jruby.jruby_puppet_agents$eval7554$send_agent__7559$fn__7560$agent_fn__7561.invoke (jruby_puppet_agents.clj:32)
    clojure.core$binding_conveyor_fn$fn__4444.invoke (core.clj:1919)
    clojure.lang.AFn.applyToHelper (AFn.java:154)
    clojure.lang.RestFn.applyTo (RestFn.java:132)
    clojure.lang.Agent$Action.doRun (Agent.java:114)
    clojure.lang.Agent$Action.run (Agent.java:163)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1145)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:615)
    java.lang.Thread.run (Thread.java:745)

Caused by: java.lang.Exception: 42
 at puppetlabs.services.jruby.jruby_puppet_service_test$fn__27808$fn__27811$fn__27812$fn__27815$fn__27816.doInvoke (jruby_puppet_service_test.clj:54)
    clojure.lang.RestFn.invoke (RestFn.java:482)
    puppetlabs.services.jruby.jruby_puppet_agents$eval7581$prime_pool_BANG___7586$fn__7590.invoke (jruby_puppet_agents.clj:53)
    puppetlabs.services.jruby.jruby_puppet_agents$eval7581$prime_pool_BANG___7586.invoke (jruby_puppet_agents.clj:38) puppetlabs.services.jruby.jruby_puppet_agents$eval7779$send_prime_pool_BANG___7784$fn__7785$fn__7787.invoke (jruby_puppet_agents.clj:181)
    puppetlabs.trapperkeeper.internal$shutdown_on_error_STAR_.invoke (internal.clj:254)
    puppetlabs.trapperkeeper.internal$shutdown_on_error_STAR_.invoke (internal.clj:238)
puppetlabs.trapperkeeper.internal$shutdown_service$reify__6433$service_fnk__5270__auto___positional$reify__6438.shutdown_on_error (internal.clj:293)
    puppetlabs.trapperkeeper.internal$eval6369$fn__6381$G__6361__6389.invoke (internal.clj:262)
    puppetlabs.trapperkeeper.internal$eval6369$fn__6381$G__6360__6398.invoke (internal.clj:262)
    clojure.core$partial$fn__4527.invoke (core.clj:2494)
    clojure.core$partial$fn__4527.invoke (core.clj:2493)
    puppetlabs.services.jruby.jruby_puppet_agents$eval7554$send_agent__7559$fn__7560$agent_fn__7561.invoke (jruby_puppet_agents.clj:32)
    clojure.core$binding_conveyor_fn$fn__4444.invoke (core.clj:1919)
    clojure.lang.AFn.applyToHelper (AFn.java:154)
    clojure.lang.RestFn.applyTo (RestFn.java:132)
    clojure.lang.Agent$Action.doRun (Agent.java:114)
    clojure.lang.Agent$Action.run (Agent.java:163)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1145)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:615)
    java.lang.Thread.run (Thread.java:745)
~~~~

It looks like this is saying that the exception that this test threw during pool initialization ended up not getting caught anywhere and, therefore, caused the test to fail.  I don't have a good theory as to why that could happen.  Seems  the exception "should" have been caught by the `shutdown_on_error` from Trapperkeeper which wraps it.

One thing that I discovered in looking into this a bit further is that several of the `jruby_puppet_service_tests` don't protect themselves against the pool agent initialization that they triggered from `with-app-with-config` calls continuing on after the test has finished.  Not sure if the possibility of multiple pool initializations running concurrently, on an intermittent basis, might somehow trigger this problem.  In any event, I added another commit to this PR, 68db764, to help better protect against that problem.  At the very least, it seems like this might help alleviate the periodic "IllegalStateException: Attempting to flush a pool that does not appear to have successfully initialized.  Aborting." messages that come up when this block of tests run and just more generally protect against other random test failures that might occur.

Ping @haus on this point, since I think he's been seeing this recently as well.

cprice404 commented:
Anything that we seems to be due to the pool initialization background thread issue I think we should try to avoid burning too much time on until we handle the ticket that specifically targets that shortcoming in the production code.

cprice404 commented:
+1 pending @camlow325 's response to my clarifying question.

cprice404 commented:
+1, would like to get this merged ASAP... @KevinCorcoran ?

haus commented:
I'm :+1: here, at least on the jruby test changes.

camlow325 commented:
Hmmm, no dice on fixing the `test-error-during-init` test failure with the latest commit.  Still failed the same way on one of the cells - https://travis-ci.org/puppetlabs/puppet-server/jobs/104130869.  I don't have any better ideas on how to address this one at this point but the other changes currently in the PR still same valuable independent of that value.

camlow325 commented:
Ok, I think found a more likely culprit for the `test-error-during-init` fail.  I put up a new commit, 34f6fe4, which I think might address it.

Basically, it's possible that if the exception during the `create-pool-instance!` were to be thrown and a shutdown requested before we get to the `run-app` call, the exception could bubble up and kill the test.  See https://github.com/puppetlabs/trapperkeeper/blob/1.1.3/test/puppetlabs/trapperkeeper/testutils/bootstrap.clj#L16-L17 in `with-app-with-config`.  I moved the try/catch to surround the whole `with-app-with-config` call.  (I think it might have been there originally before I stupidly suggested to @haus in a recent PR that he change it).

Will be interesting to see if Travis runs clean this time...

camlow325 commented:
Last Travis run was green - https://travis-ci.org/puppetlabs/puppet-server/builds/104153447.  We good to ship this, now?  @cprice404?

camlow325 commented:
er, I just rekicked Travis one more time on the latest code to see if all of the OSes will pass again in the meantime.  Will leave it to others discretion as to whether or not we want to wait for the job to complete again before merging.

camlow325 commented:
Travis passed a second time on all cells with the latest PR changes.  Same link as above.

camlow325 commented:
I noticed a Travis fail from #874 that seemed related to the fixes on this PR.  Here was that fail:

~~~~
lein test :only puppetlabs.services.jruby.jruby-locking-test/with-lock-event-notification-test

ERROR in (with-lock-event-notification-test) (jruby_locking_test.clj:83)
Uncaught exception, not in assertion.
expected: nil

  actual: clojure.lang.ExceptionInfo: throw+: {:type :puppetlabs.services.jruby.jruby-puppet-service/jruby-timeout, :message "Attempt to borrow a JRuby instance from the pool timed out; Puppet Server is temporarily overloaded. If you get this error repeatedly, your server might be misconfigured or trying to serve too many agent nodes. Check Puppet Server settings: jruby-puppet.max-active-instances."}

 at puppetlabs.services.jruby.jruby_locking_test$fn__25014$fn__25058$fn__25059.invoke (jruby_locking_test.clj:83)
    puppetlabs.services.jruby.jruby_locking_test$fn__25014$fn__25058.invoke (jruby_locking_test.clj:82)
    clojure.core$with_redefs_fn.invoke (core.clj:7209)
    puppetlabs.services.jruby.jruby_locking_test/fn (jruby_locking_test.clj:74)
    clojure.test$test_var$fn__7670.invoke (test.clj:704)
    clojure.test$test_var.invoke (test.clj:704)
    clojure.test$test_vars$fn__7692$fn__7697.invoke (test.clj:722)
    puppetlabs.services.jruby.jruby_testutils$mock_pool_instance_fixture$fn__8818.invoke (jruby_testutils.clj:170)
    clojure.core$with_redefs_fn.invoke (core.clj:7209)
    puppetlabs.services.jruby.jruby_testutils$mock_pool_instance_fixture.invoke (jruby_testutils.clj:168)
    clojure.test$compose_fixtures$fn__7664$fn__7665.invoke (test.clj:681)
    clojure.test$default_fixture.invoke (test.clj:674)
    clojure.test$compose_fixtures$fn__7664.invoke (test.clj:681)
    clojure.test$test_vars$fn__7692.invoke (test.clj:722)
    schema.test$validate_schemas$fn10790__10799$fn__10800.invoke (test.clj:11)
    schema.test$validate_schemas$fn10790__10799.invoke (test.clj:11)
    clojure.lang.AFn.applyToHelper (AFn.java:152)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.invoke (RestFn.java:397)
    schema.test$validate_schemas.invoke (test.clj:11)
    clojure.test$compose_fixtures$fn__7664$fn__7665.invoke (test.clj:681)
    clojure.test$default_fixture.invoke (test.clj:674)
    clojure.test$compose_fixtures$fn__7664.invoke (test.clj:681)
    clojure.test$test_vars.invoke (test.clj:718)
    clojure.test$test_all_vars.invoke (test.clj:728)
    clojure.test$test_ns.invoke (test.clj:747)
    clojure.core$map$fn__4553.invoke (core.clj:2624)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.next (RT.java:674)
    clojure.core/next (core.clj:64)
    clojure.core$reduce1.invoke (core.clj:909)
    clojure.core$reduce1.invoke (core.clj:900)
    clojure.core$merge_with.doInvoke (core.clj:2936)
    clojure.lang.RestFn.applyTo (RestFn.java:139)
    clojure.core$apply.invoke (core.clj:632)
    clojure.test$run_tests.doInvoke (test.clj:762)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invoke (core.clj:630)
    user$eval923$fn__978$fn__1009.invoke (form-init4862221704814442006.clj:1)
    user$eval923$fn__978$fn__979.invoke (form-init4862221704814442006.clj:1)
    user$eval923$fn__978.invoke (form-init4862221704814442006.clj:1)
    user$eval923.invoke (form-init4862221704814442006.clj:1)
    clojure.lang.Compiler.eval (Compiler.java:6782)
    clojure.lang.Compiler.eval (Compiler.java:6772)
    clojure.lang.Compiler.load (Compiler.java:7227)
    clojure.lang.Compiler.loadFile (Compiler.java:7165)
    clojure.main$load_script.invoke (main.clj:275)
    clojure.main$init_opt.invoke (main.clj:280)
    clojure.main$initialize.invoke (main.clj:308)
    clojure.main$null_opt.invoke (main.clj:343)
    clojure.main$main.doInvoke (main.clj:421)
    clojure.lang.RestFn.invoke (RestFn.java:421)
    clojure.lang.Var.invoke (Var.java:383)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.main.main (main.java:37)
~~~~

This appeared to be due to the very short borrow timeout that the test uses, 1 millisecond, coupled with the fact that the test does not wait until an instance is available in the pool to be borrowed before attempting the borrow.  Like with some of the other tests in this PR, I added another `wait-for-jrubies` call to this test to close up the race condition.  I added that in 4846f60.
